### DISCUSSION TITLE
C++17 support

### DISCUSSION THREAD

**Discussion by nullsysvoid** [Soundux/rohrkabel#6] [Q&A] [answered]
Hi! I would like to use this library with embedded device but it uses arm gcc compiler with C++17 support and there is no any plans to move to new C++20 standard. So I would like to know if you have any plans to use C++17 in this project? I looked at the code and in general only concepts and format are used from the new standard.

**Comment by Curve** [comment, marked as answer]
I have C++17 support in versions [1.5](https://github.com/Curve/rohrkabel/releases/tag/v1.5) and prior.

As for C++17 support in the current releases of the library I'm not sure if I want to do that - Sure the features used from C++20 are few but back porting them with conditional implementations and macros is not something I'm fond of doing.

**However**: If I find the time to do so, I can create a C++17 branch (with separate tags) and back port the current implementation there. I can't guarantee that the C++17 branch will always be up to date then but I'll do my best to keep them in sync.

**Reply by nullsysvoid** [reply]
Thank you

**Reply by nullsysvoid** [reply]
Also
Macros do not have to be used inside the source code, you can make 2 implementations and link them depending on the standard in cmake. This will make the code clean from conditional compilations

**Reply by Curve** [reply]
> Also
> Macros do not have to be used inside the source code, you can make 2 implementations and link them depending on the standard in cmake. This will make the code clean from conditional compilations

This is what I do in all of my other projects but this is not true for header files, where the most C++20 features are used currently.

Sure we could have conditional includes but that just makes everything unnecessarily cluttered.

**Reply by nullsysvoid** [reply]
This can also be solved, for example, instead of using fmt directly, you can use a wrapper class over it

**Reply by Curve** [reply]
> This can also be solved, for example, instead of using fmt directly, you can use a wrapper class over it

Format is not used in the headers but concepts, which are a language feature...

Also, due to the template heavy nature of format a non-header heavy approach is not plausible

**Comment by Curve** [comment]
Okay, I've just looked into this - And back-porting the current code-base is a lot more work than I initially anticipated due to some of the dependencies also relying on C++20.

I don't think it's currently in scope to add C++17 support back in completely.  
However using [Rohrkabel v1.5](https://github.com/Curve/rohrkabel/releases/tag/v1.5) should work fine since most updates since then have mostly been QoL changes.

**Reply by nullsysvoid** [reply]
Thank you for response. I will try to check with ver.1.5
